packages/dd-trace/src/llmobs/sdk.js

```diff
@@ -0,0 +1,448 @@
+        if (this.enabled) storage.enterWith(oldStore)
+        throw e
+      }
+    })
+  }
+
+  wrap (options = {}, fn) {
+    if (typeof options === 'function') {
+      fn = options
+      options = {}
+    }
+
+    const kind = options.kind
+    const name = getName(kind, options, fn)
+
+    const {
+      spanOptions,
+      ...llmobsOptions
+    } = this._extractOptions(options)
+
+    const llmobs = this
+
+    function wrapped () {
+      if (!llmobs.enabled) {
+        logger.warn('Span started while LLMObs is disabled. Spans will not be sent to LLM Observability.')
+      }
+
+      const valid = validKind(kind)
+      if (!valid) {
+        logger.warn(`Invalid span kind specified: ${kind}. Span will not be sent to LLM Observability.`)
+      }
+
+      const span = llmobs._tracer.scope().active()
+      const oldStore = storage.getStore()
+      const parentLLMObsSpan = oldStore?.llmobsSpan
+      if (llmobs.enabled) storage.enterWith({ ...oldStore, llmobsSpan: span })
+
+      llmobs._tagger.setLLMObsSpanTags(span, valid && kind, {
+        ...llmobsOptions,
+        parentLLMObsSpan
+      })
+      llmobs.annotate(span, { inputData: getFunctionArguments(fn, arguments) })
+
+      try {
+        const result = fn.apply(this, arguments)
+
+        if (result && typeof result.then === 'function') {
+          return result.then(value => {
+            if (value && kind !== 'retrieval' && !LLMObsTagger.tagMap.get(span)?.[OUTPUT_VALUE]) {
+              llmobs.annotate(span, { outputData: value })
+            }
+            if (llmobs.enabled) storage.enterWith(oldStore)
+            return value
+          }).catch(err => {
+            if (llmobs.enabled) storage.enterWith(oldStore)
+            throw err
+          })
+        }
+
+        if (result && kind !== 'retrieval' && !LLMObsTagger.tagMap.get(span)?.[OUTPUT_VALUE]) {
+          llmobs.annotate(span, { outputData: result })
+          if (llmobs.enabled) storage.enterWith(oldStore)
+        }
+
+        return result
+      } catch (e) {
+        if (llmobs.enabled) storage.enterWith(oldStore)
+        throw e
+      }
+    }
+
+    return this._tracer.wrap(name, spanOptions, wrapped)
+  }
+
+  annotate (span, options) {
+    if (!this.enabled) {
+      logger.warn(
+        'Annotate called while LLMObs is disabled. Not annotating span.'
+      )
+      return
+    }
+
+    if (!span) {
+      span = this.active()
+    }
+
+    if ((span && !options) && !(span instanceof Span)) {
+      options = span
+      span = this.active()
+    }
+
+    if (!span) {
+      logger.warn('No span provided and no active LLMObs-generated span found')
+      return
+    }
+    if (!options) {
+      logger.warn('No options provided for annotation.')
+      return
+    }
+
+    if (!LLMObsTagger.tagMap.has(span)) {
+      logger.warn('Span must be an LLMObs-generated span')
+      return
+    }
+    if (span._duration !== undefined) {
+      logger.warn('Cannot annotate a finished span')
```

Comment:
all these above cases basically mean that the annotation failed and we're failing silently resulting in a poor product experience. I think we should throw an exception for each of these.

When doing auto instrumentation we should opt more for soft fails since the user isn't agreeing to an explicit contract in their code.